### DISCUSSION TITLE
How to define a type of class implementing an interface

### DISCUSSION THREAD

**Discussion by joostvdwsd** [aws/jsii#3463] [Q&A]
### Affected Languages

- [X] `TypeScript` or `Javascript`
- [ ] `Python`
- [ ] `Java`
- [ ] .NET (`C#`, `F#`, ...)
- [ ] `Go`

### General Information
* **JSII Version:** 1.30.0 (build adae23f), typescript 3.9.9
* **Platform:**     macos

### The Question

I like to define a type that is a subclass of class x and follows interface Y. This is a pretty standard oo pattern but I can't seem to find a way where the JSII compiler will be happy.

### Attempt 1 (Works in typescript but not jsii)

```typescript
export class Foo {
  bar: string
}

export interface IActivatable {
  activate();
}

export interface ActiveFoo extends Foo, IActivatable {
}

function test(activeFoo: ActiveFoo) {
}

```
This will give the jsii error: "Type "Foo" cannot be used as an interface"

### Attempt 2 (Works in typescript but not jsii)

```typescript
export class Foo {
 bar: string
}

export interface IActivatable {
  activate();
}

export type ActiveFoo = Foo & IActivatable

function test(activeFoo: ActiveFoo) {
}

```
The above example gives the jsii error: "Non-primitive types without a symbol cannot be processed."

Also attempt # 2 but with the type definitions in generics will give the same error.

### Attempt 3 (Works but looses type enforcement)

```typescript
export class Foo {
 bar: string
}

export interface IActivatable {
  activate();
}

export function isActivatable(arg: any) : arg is IActivatable {
  return (arg as IActivatable).activate !== undefined;
}

function test(activeFoo: Foo) {
  if (isActivatable(activeFoo)) {
  }
}

```
This last one looses the type enforcement and its not clear that it will check for the interface type. 

# Question
Is there any way to achieve the type enforcement in the test function for jsii?

**Comment by webratz** [comment]
Did you ever find a solution for this?

**Comment by chub-jancel** [comment]
I'd like to hear more on this topic and the other references floating around for this particular issue. Is the third option what most are going with and is there a gist of a generic approach? How did you solve it?
